File: aoc24/python/day02.py
with open("aoc24/inputs/day02.txt") as f:
    lines = f.read().splitlines()

reports = [[int(i) for i in line.split(" ")] for line in lines]


def is_safe(report: list[int]) -> bool:
    diffs = {report[i] - report[i + 1] for i in range(len(report) - 1)}
    return diffs.issubset({1, 2, 3}) or diffs.issubset({-1, -2, -3})


def dampened_is_safe(report: list[int]) -> bool:
    if is_safe(report):
        return True

    for i in range(len(report)):
        copy = report[:]
        copy.pop(i)
        if is_safe(copy):
            return True

    return False


part_1 = len([report for report in reports if is_safe(report)])
part_2 = len([report for report in reports if dampened_is_safe(report)])

print(f"Part 1: {part_1}")
print(f"Part 2: {part_2}")

File: aoc24/src/day02.clj
(ns aoc24.src.day02
  (:require
   [aoc-tools :refer [read-input submit-answer]]
   [clojure.string :as str]
   [utils :refer [parse-int nmap]]))

(def input (read-input :test))
(def real-input (read-input))

(defn safe?
  [coll]
  (let [diffs (map - coll (rest coll))]
    (or
     (every? #{1 2 3} diffs)
     (every? #{-1 -2 -3} diffs))))

;; Part 1
(->> real-input
     (map #(str/split % #" "))
     (nmap 2 parse-int)
     (filter safe?)
     (count)
     #_(submit-answer 1))

(defn drop-nth
  [n coll]
  (concat
   (take n coll)
   (drop (inc n) coll)))

(defn dampened-safe?
  [coll]
  (if (safe? coll)
    true
    (->> coll
         (count)
         (range)
         (map #(drop-nth % coll))
         (some safe?))))

;; Part 2
(->> real-input
     (map #(str/split % #" "))
     (nmap 2 parse-int)
     (filter dampened-safe?)
     (count)
     #_(submit-answer 2))
